zkevm-circuits/src/state_circuit/lexicographic_ordering.rs

```diff
@@ -1,220 +1,187 @@
-            rw_counter_limbs,
+            first_different_limb,
+            limb_difference,
+            limb_difference_inverse,
         };
-        meta.create_gate("upper_limb_difference is one of 15 values", |meta| {
-            let selector = meta.query_fixed(selector, Rotation::cur());
-            let cur = Queries::new(meta, &config, Rotation::cur());
-            let prev = Queries::new(meta, &config, Rotation::prev());
-            let upper_limb_difference = meta.query_advice(upper_limb_difference, Rotation::cur());
-            vec![
-                selector
-                    * upper_limb_difference_possible_values(cur, prev)
-                        .iter()
-                        .map(|e| upper_limb_difference.clone() - e.clone())
-                        .fold(1.expr(), Expression::mul),
-            ]
-        });
-        assert!(meta.degree() <= 16);
-        meta.create_gate(
-            "upper_limb_difference is zero iff all 15 possible values are 0",
-            |meta| {
-                let selector = meta.query_fixed(selector, Rotation::cur());
-                let cur = Queries::new(meta, &config, Rotation::cur());
-                let prev = Queries::new(meta, &config, Rotation::prev());
-                vec![
-                    // all 15 possible values are 0 iff the final linear combination is 0
-                    selector
-                        * upper_limb_difference_is_zero.clone()
-                        * upper_limb_difference_possible_values(cur, prev)[14].clone(),
-                ]
-            },
-        );
-        meta.create_gate("lower_limb_difference is one of 15 values", |meta| {
-            let selector = meta.query_fixed(selector, Rotation::cur());
-            let cur = Queries::new(meta, &config, Rotation::cur());
-            let prev = Queries::new(meta, &config, Rotation::prev());
-            let lower_limb_difference = meta.query_advice(lower_limb_difference, Rotation::cur());
-            vec![
-                selector
-                    * lower_limb_difference_possible_values(cur, prev)
-                        .iter()
-                        .map(|e| lower_limb_difference.clone() - e.clone())
-                        .fold(1.expr(), Expression::mul),
-            ]
-        });
-        assert!(meta.degree() <= 16);
-        meta.lookup_any("upper_limb_difference fits into u16", |meta| {
-            let upper_limb_difference = meta.query_advice(upper_limb_difference, Rotation::cur());
+
+        meta.lookup_any("limb_difference fits into u16", |meta| {
             vec![(
-                upper_limb_difference,
+                meta.query_advice(limb_difference, Rotation::cur()),
                 meta.query_fixed(u16_range, Rotation::cur()),
             )]
         });
-        meta.lookup_any(
-            "upper_limb_difference is zero or lower_limb_difference fits into u16",
+
+        meta.create_gate("limb_difference is not zero", |meta| {
```

Comment:
It would be nice to have an IsZeroGadget that doesn't depend on the evm_circuit ConstraintBuilder!

Reply:
`IsZeroGadget` doesn't depend on the evm `ConstraintBuilder`, but not using it saves one constraint here.

Reply:
Oh I was thinking about this [IsZeroGadget](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/d567bb23fe16eedfb15ff8a67063de201724ae49/zkevm-circuits/src/evm_circuit/util/math_gadget.rs#L20) but I just saw that we also have an `IsZeroChip` under `gadgets`.  I now see the one saved constraint, thanks!